File: src/utils/toml.zig
const std = @import("std");
const c = @cImport({
    @cInclude("toml_parser/parser.h");
});

pub const ValueType = enum {
    integer,
    float_val,
    string,
    bool_val,
    datetime,
    array,
    inline_table,
    null,
};

pub const Array = struct {
    items: []const Value,
    allocator: std.mem.Allocator,

    pub fn len(self: Array) usize {
        return self.items.len;
    }

    pub fn get(self: Array, index: usize) ?Value {
        if (index >= self.items.len) return null;
        return Value{ .inner = self.items[index] };
    }

    pub fn string(self: Array, index: usize) ?[]const u8 {
        const v = self.get(index) orelse return null;
        return v.string();
    }

    pub fn int(self: Array, index: usize) ?i64 {
        const v = self.get(index) orelse return null;
        return v.int();
    }

    pub fn table(self: Array, index: usize) ?Table {
        const v = self.get(index) orelse return null;
        return v.table() orelse return null;
    }
};

pub const Value = struct {
    inner: c.Value,

    pub fn typeOf(self: Value) ValueType {
        return switch (self.inner.type) {
            c.VAL_INTEGER => .integer,
            c.VAL_FLOAT => .float_val,
            c.VAL_STRING => .string,
            c.VAL_BOOL => .bool_val,
            c.VAL_DATETIME => .datetime,
            c.VAL_ARRAY => .array,
            c.VAL_INLINE_TABLE => .inline_table,
            else => .null,
        };
    }

    pub fn string(self: Value) ?[]const u8 {
        if (self.inner.type != c.VAL_STRING) return null;
        const ptr = self.inner.as.string;
        if (ptr == null) return null;
        return std.mem.sliceTo(ptr, 0);
    }

    pub fn int(self: Value) ?i64 {
        if (self.inner.type != c.VAL_INTEGER) return null;
        return @intCast(self.inner.as.integer);
    }

    pub fn float(self: Value) ?f64 {
        if (self.inner.type != c.VAL_FLOAT) return null;
        return self.inner.as.float_val;
    }

    pub fn bool(self: Value) ?bool {
        if (self.inner.type != c.VAL_BOOL) return null;
        return self.inner.as.boolean;
    }

    pub fn datetime(self: Value) ?[]const u8 {
        if (self.inner.type != c.VAL_DATETIME) return null;
        const ptr = self.inner.as.datetime;
        if (ptr == null) return null;
        return std.mem.sliceTo(ptr, 0);
    }

    pub fn array(self: Value) ?Array {
        if (self.inner.type != c.VAL_ARRAY) return null;
        const arr = self.inner.as.array;
        if (arr.items == null or arr.len == 0) return .{
            .items = &.{},
            .allocator = undefined,
        };
        return .{
            .items = @as([*]const c.Value, @ptrCast(arr.items))[0..arr.len],
            .allocator = undefined,
        };
    }

    pub fn table(self: Value) ?Table {
        if (self.inner.type != c.VAL_INLINE_TABLE) return null;
        const ptr = self.inner.as.inline_table;
        if (ptr == null) return null;
        return .{ .inner = ptr };
    }
};

pub const Table = struct {
    inner: *c.Table,

    pub fn get(self: Table, key: []const u8) ?Value {
        var i: usize = 0;
        while (i < self.inner.entry_count) : (i += 1) {
            const entry = self.inner.entries[i];
            if (entry.key.segment_count == 1) {
                const seg = entry.key.segments[0];
                const seg_len = std.mem.sliceTo(seg, 0);
                if (std.mem.eql(u8, seg_len, key)) {
                    return .{ .inner = entry.value };
                }
            }
        }
        return null;
    }

    pub fn string(self: Table, key: []const u8) ?[]const u8 {
        const v = self.get(key) orelse return null;
        return v.string();
    }

    pub fn int(self: Table, key: []const u8) ?i64 {
        const v = self.get(key) orelse return null;
        return v.int();
    }

    pub fn float(self: Table, key: []const u8) ?f64 {
        const v = self.get(key) orelse return null;
        return v.float();
    }

    pub fn bool(self: Table, key: []const u8) ?bool {
        const v = self.get(key) orelse return null;
        return v.bool();
    }

    pub fn datetime(self: Table, key: []const u8) ?[]const u8 {
        const v = self.get(key) orelse return null;
        return v.datetime();
    }

    pub fn array(self: Table, key: []const u8) ?Array {
        const v = self.get(key) orelse return null;
        return v.array();
    }

    pub fn table(self: Table, key: []const u8) ?Table {
        var i: usize = 0;
        while (i < self.inner.child_count) : (i += 1) {
            const child = self.inner.children[i];
            if (child.name == null) continue;
            const name = std.mem.sliceTo(child.name, 0);
            if (std.mem.eql(u8, name, key)) {
                return .{ .inner = child };
            }
        }
        return null;
    }
};

pub const ParsedToml = struct {
    program: *c.Program,
    lexer: c.Lexer,
    allocator: std.mem.Allocator,
    owned: []u8,

    pub fn parse(allocator: std.mem.Allocator, file_path: []const u8) !ParsedToml {
        var file = try std.fs.cwd().openFile(file_path, .{});
        defer file.close();

        const stat = try file.stat();
        const size: usize = @intCast(stat.size);

        var owned = try allocator.alloc(u8, size + 1);
        const n = try file.readAll(owned);
        owned[n] = 0;
        const content = owned[0..n :0];

        var lexer: c.Lexer = undefined;
        c.lexer_init(&lexer, content.ptr, content.len);
        if (!c.lexer_tokenize(&lexer)) {
            return error.LexError;
        }

        const program = c.program_init();
        errdefer {
            // program is heap-allocated by C, would need a program_destroy
            allocator.free(owned);
        }

        var parser = c.parser_init(&lexer);
        if (!c.parser_parse(&parser, program)) {
            return error.ParseError;
        }

        return .{
            .program = program,
            .lexer = lexer,
            .allocator = allocator,
            .owned = owned,
        };
    }

    pub fn deinit(self: *ParsedToml) void {
        self.allocator.free(self.owned);
    }

    pub fn root(self: *const ParsedToml) Table {
        return .{ .inner = &self.program.root };
    }

    pub fn table(self: *const ParsedToml, name: []const u8) ?Table {
        return self.root().table(name);
    }

    pub fn string(self: *const ParsedToml, key: []const u8) ?[]const u8 {
        return self.root().string(key);
    }

    pub fn int(self: *const ParsedToml, key: []const u8) ?i64 {
        return self.root().int(key);
    }

    pub fn bool(self: *const ParsedToml, key: []const u8) ?bool {
        return self.root().bool(key);
    }
};

File: src/utils/toml_parser/toml.cpp
#include "toml.hpp"

namespace Pico::TOML {

Toml::Toml(const std::string &src) : lexer_(src), parser_(lexer_) {
    if (!lexer_.Lexer_tokenize()) {
        ok_ = false;
        error_msg_ = lexer_.error_msg;
        error_line_ = lexer_.error_line;
        error_col_ = lexer_.error_col;
        return;
    }
    if (!parser_.parse(&program_)) {
        ok_ = false;
        error_msg_ = parser_.error_msg;
        error_line_ = parser_.error_line;
        error_col_ = parser_.error_col;
        return;
    }
    ok_ = true;
}

Toml::~Toml() = default;

Node Toml::node(const std::string &key) const {
    return Table(&program_.root).node(key);
}

Table Toml::table(const std::string &name) const {
    for (TomlTable *c : program_.root.children)
        if (!c->name.empty() && c->name == name)
            return Table(c);
    return Table();
}

std::vector<Table> Toml::tables(const std::string &name) const {
    std::vector<Table> out;
    for (TomlTable *c : program_.root.children)
        if (!c->name.empty() && c->name == name)
            out.push_back(Table(c));
    return out;
}

Node Table::node(const std::string &key) const {
    if (!table_)
        return Node();

    size_t dot = key.find('.');
    if (dot != std::string::npos)
        return table(key.substr(0, dot)).node(key.substr(dot + 1));

    for (const Entry &e : table_->entries)
        if (!e.key.segments.empty() && e.key.segments[0] == key)
            return Node(&e.value);
    return Node();
}

Table Table::table(const std::string &name) const {
    if (!table_)
        return Table();
    for (TomlTable *c : table_->children)
        if (!c->name.empty() && c->name == name)
            return Table(c);
    return Table();
}

bool Table::contains(const std::string &key) const { return node(key).valid(); }

std::string Node::as_string() const {
    if (!value_ || value_->type != ValType::VAL_STRING)
        throw TomlError("expected TOML string");
    return std::get<std::string>(value_->as);
}

long Node::as_integer() const {
    if (!value_)
        throw TomlError("expected TOML integer");
    if (value_->type == ValType::VAL_FLOAT)
        return static_cast<long>(std::get<double>(value_->as));
    if (value_->type != ValType::VAL_INTEGER)
        throw TomlError("expected TOML integer");
    return std::get<long>(value_->as);
}

double Node::as_float() const {
    if (!value_)
        throw TomlError("expected TOML float");
    if (value_->type == ValType::VAL_INTEGER)
        return static_cast<double>(std::get<long>(value_->as));
    if (value_->type != ValType::VAL_FLOAT)
        throw TomlError("expected TOML float");
    return std::get<double>(value_->as);
}

bool Node::as_bool() const {
    if (!value_ || value_->type != ValType::VAL_BOOL)
        throw TomlError("expected TOML bool");
    return std::get<bool>(value_->as);
}

Table Node::as_table() const {
    if (!value_ || value_->type != ValType::VAL_INLINE_TABLE)
        throw TomlError("expected TOML table");
    return Table(std::get<TomlTable *>(value_->as));
}

size_t Node::size() const {
    if (!value_ || value_->type != ValType::VAL_ARRAY)
        throw TomlError("expected TOML array");
    return std::get<std::vector<Value>>(value_->as).size();
}

Node Node::operator[](size_t i) const {
    if (!value_ || value_->type != ValType::VAL_ARRAY)
        throw TomlError("expected TOML array");
    const std::vector<Value> &arr = std::get<std::vector<Value>>(value_->as);
    if (i >= arr.size())
        throw TomlError("array index out of bounds");
    return Node(&arr[i]);
}

} // namespace Pico::TOML
